From b37c98ddd1b8cd471d02d48102f0f917f271ea0a Mon Sep 17 00:00:00 2001
From: Vitaly Osipov <vitaly.osipov@gmail.com>
Date: Sat, 12 Apr 2014 21:25:52 +1000
Subject: [PATCH] Adding id sysfs attribute to FTDI serial driver

A new attribute, only for FT232RL FTDI chips. Functionality ported from the earlier sysfs driver task.

Signed-off-by: Vitaly Osipov <vitaly.osipov@gmail.com>
---
vosipov@ubuntu:~/kernel/eudyptula/task11$ ls -la /sys/devices/pci0000:00/0000:00:11.0/0000:02:00.0/usb2/2-2/2-2.1/2-2.1:1.0/ttyUSB0
total 0
drwxr-xr-x 4 root root    0 Apr 12 21:17 .
drwxr-xr-x 6 root root    0 Apr 12 21:17 ..
lrwxrwxrwx 1 root root    0 Apr 12 21:17 driver -> ../../../../../../../../../bus/usb-serial/drivers/ftdi_sio
--w------- 1 root root 4096 Apr 12 21:17 event_char
-rw-rw-rw- 1 root root 4096 Apr 12 21:17 id
-rw-r--r-- 1 root root 4096 Apr 12 21:17 latency_timer
-r--r--r-- 1 root root 4096 Apr 12 21:17 port_number
drwxr-xr-x 2 root root    0 Apr 12 21:17 power
lrwxrwxrwx 1 root root    0 Apr 12 21:17 subsystem -> ../../../../../../../../../bus/usb-serial
drwxr-xr-x 3 root root    0 Apr 12 21:17 tty
-rw-r--r-- 1 root root 4096 Apr 12 21:17 uevent
vosipov@ubuntu:~/kernel/eudyptula/task11$ cat /sys/devices/pci0000:00/0000:00:11.0/0000:02:00.0/usb2/2-2/2-2.1/2-2.1:1.0/ttyUSB0/id
7c1caf2f50d1
vosipov@ubuntu:~/kernel/eudyptula/task11$ echo "7c1caf2f50d1" > /sys/devices/pci0000:00/0000:00:11.0/0000:02:00.0/usb2/2-2/2-2.1/2-2.1:1.0/ttyUSB0/id
vosipov@ubuntu:~/kernel/eudyptula/task11$ echo "blah" > /sys/devices/pci0000:00/0000:00:11.0/0000:02:00.0/usb2/2-2/2-2.1/2-2.1:1.0/ttyUSB0/id
bash: echo: write error: Invalid argument
vosipov@ubuntu:~/kernel/eudyptula/task11$

 drivers/usb/serial/ftdi_sio.c |   29 +++++++++++++++++++++++++++++
 1 file changed, 29 insertions(+)

diff --git a/drivers/usb/serial/ftdi_sio.c b/drivers/usb/serial/ftdi_sio.c
index 44ab129..3785076 100644
--- a/drivers/usb/serial/ftdi_sio.c
+++ b/drivers/usb/serial/ftdi_sio.c
@@ -50,6 +50,9 @@
 #define DRIVER_AUTHOR "Greg Kroah-Hartman <greg@kroah.com>, Bill Ryder <bryder@sgi.com>, Kuba Ober <kuba@mareimbrium.org>, Andreas Mohr, Johan Hovold <jhovold@gmail.com>"
 #define DRIVER_DESC "USB FTDI Serial Converters Driver"
 
+#define MY_ID "7c1caf2f50d1\n"
+#define MY_ID_LEN 13		/* MY_ID length */
+
 
 struct ftdi_private {
 	enum ftdi_chip_type chip_type;
@@ -1623,6 +1626,28 @@ static ssize_t store_event_char(struct device *dev,
 }
 static DEVICE_ATTR(event_char, S_IWUSR, NULL, store_event_char);
 
+static ssize_t id_show(struct device *dev, struct device_attribute *attr,
+		       char *buf)
+{
+	char *hello_str = MY_ID;
+
+	strncpy(buf, hello_str, MY_ID_LEN);
+	return MY_ID_LEN;
+}
+
+static ssize_t id_store(struct device *dev, struct device_attribute *attr,
+			const char *buf, size_t count)
+{
+	char *hello_str = MY_ID;
+
+	if ((count != MY_ID_LEN) || (strncmp(hello_str, buf, MY_ID_LEN)))
+		return -EINVAL;
+	else
+		return MY_ID_LEN;
+}
+
+static DEVICE_ATTR(id, 0666, id_show, id_store);
+
 static int create_sysfs_attrs(struct usb_serial_port *port)
 {
 	struct ftdi_private *priv = usb_get_serial_port_data(port);
@@ -1645,6 +1670,10 @@ static int create_sysfs_attrs(struct usb_serial_port *port)
 						    &dev_attr_latency_timer);
 		}
 	}
+
+	if ((!retval) && (priv->chip_type == FT232RL))
+		retval = device_create_file(&port->dev, &dev_attr_id);
+
 	return retval;
 }
 
-- 
1.7.9.5

